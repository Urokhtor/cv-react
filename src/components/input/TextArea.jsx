import React, {Component} from 'react';
import PropTypes from 'prop-types';
import './TextArea.scss';

const DEFAULT_WIDTH = 250;
const DEFAULT_HEIGHT = 70;

export default class TextArea extends Component {

    constructor(props) {
        super(props);

        this.state = {
            value: props.value,
            error: null
        };
    }

    static getDerivedStateFromProps(nextProps, prevState) {
        if (prevState.value !== nextProps.value) {
            return {
                value: nextProps.value
            };
        }

        return null;
    }

    onChange(event) {
        this.setState({
            value: event.target.value
        });
    }

    onBlur() {
        this.setState({
            error: null
        });

        let failed = false;

        if (this.props.validator) {
            let validationResult = this.props.validator(this.state.value);

            if (validationResult !== undefined) {
                failed = true;
                this.setState({
                    error: validationResult
                });
            }
        }

        this.props.onChange({
            failed: failed,
            value: this.state.value
        });
    }

    render () {
        return (
            <div className={"text-area flex-vertical"}>
                <textarea
                    style={{
                        width: this.props.width,
                        height: this.props.height
                    }}
                    onChange={this.onChange.bind(this)}
                    onBlur={this.onBlur.bind(this)}
                    value={this.state.value ? this.state.value : ''}
                    required/>
                {this.props.label && <label className={"label"}>{this.props.label}</label>}
                {this.state.error && <label className={"error"}>{this.state.error}</label>}
            </div>
        );
    }
}

TextArea.propTypes = {
    value: PropTypes.string,
    label: PropTypes.string,
    onChange: PropTypes.func.isRequired,
    validator: PropTypes.func,
    width: PropTypes.number,
    height: PropTypes.number
};

TextArea.defaultProps = {
    width: DEFAULT_WIDTH,
    height: DEFAULT_HEIGHT
};