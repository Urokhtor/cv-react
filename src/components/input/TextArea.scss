@import '../../styles/index';

.text-area {
  textarea {
    background: transparent;

    transition: 0.2s border ease-in-out;

    resize: none;
    outline: none;
    border: 0;
    border-bottom: 2px solid $colour-secondary;
    padding-top: 16px;
  }

  textarea:focus {
    border-bottom-color: $colour-accent;
  }

  textarea:focus::placeholder {
    position: absolute;
    top: 0;
    font-size: 75%;
    color: $colour-accent;
    transition: all 0.2s ease-in-out;
  }
}