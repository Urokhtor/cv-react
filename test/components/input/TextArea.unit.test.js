import React from 'react';
import TextArea from "../../../src/components/input/TextArea";

test('TextArea should match snapshot', (done) => {
    const mockOnChange = jest.fn();
    const value = 'some-text-value';
    const wrapper = shallow(<TextArea
            onChange={mockOnChange}
            label={'some-text-label'}
            value={value} />);

    expect(wrapper).toMatchSnapshot();

    done();
});

test('TextArea should match snapshot when in failed state', (done) => {
    const mockOnChange = jest.fn();
    const value = 'some-text-value';
    const wrapper = shallow(<TextArea
            onChange={mockOnChange}
            validator={() => "error"}
            label={'some-text-label'}
            value={value} />);

    const textarea = wrapper.find('textarea');
    textarea.simulate('click');
    textarea.simulate('change', { target: { value: 'new-text-input' }});
    textarea.simulate('blur');

    expect(wrapper).toMatchSnapshot();

    done();
});

test('TextArea should return current text value on blur', (done) => {
    const mockOnChange = jest.fn();
    const value = 'some-text-value';
    const wrapper = shallow(<TextArea
        onChange={mockOnChange}
        label={'some-text-label'}
        value={value}/>);

    const textarea = wrapper.find('textarea');
    textarea.simulate('click');
    textarea.simulate('change', { target: { value: 'new-text-input' }});
    textarea.simulate('blur');

    expect(mockOnChange).toHaveBeenCalledWith({
        failed: false,
        value: 'new-text-input'
    });

    done();
});

test('TextArea should show error if validation fails', (done) => {
    const mockOnChange = jest.fn();
    const value = 'some-text-value';
    const wrapper = shallow(<TextArea
        onChange={mockOnChange}
        validator={() => "error"}
        label={'some-text-label'}
        value={value}/>);

    const textarea = wrapper.find('textarea');
    textarea.simulate('click');
    textarea.simulate('change', { target: { value: 'new-text-input' }});
    textarea.simulate('blur');

    expect(mockOnChange).toHaveBeenCalledWith({
        failed: true,
        value: 'new-text-input'
    });

    expect(wrapper.find('label .error').text()).toEqual('error');

    done();
});