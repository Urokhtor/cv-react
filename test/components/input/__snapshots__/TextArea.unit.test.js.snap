// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TextArea
    height={70}
    label="some-text-label"
    onChange={[MockFunction]}
    value="some-text-value"
    width={250}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <textarea
          onBlur={[Function]}
          onChange={[Function]}
          required={true}
          style={
            Object {
              "height": 70,
              "width": 250,
            }
          }
          value="some-text-value"
        />,
        <label
          className="label"
        >
          some-text-label
        </label>,
        null,
      ],
      "className": "text-area flex-vertical",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "onBlur": [Function],
          "onChange": [Function],
          "required": true,
          "style": Object {
            "height": 70,
            "width": 250,
          },
          "value": "some-text-value",
        },
        "ref": null,
        "rendered": null,
        "type": "textarea",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "some-text-label",
          "className": "label",
        },
        "ref": null,
        "rendered": "some-text-label",
        "type": "label",
      },
      null,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <textarea
            onBlur={[Function]}
            onChange={[Function]}
            required={true}
            style={
              Object {
                "height": 70,
                "width": 250,
              }
            }
            value="some-text-value"
          />,
          <label
            className="label"
          >
            some-text-label
          </label>,
          null,
        ],
        "className": "text-area flex-vertical",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "onBlur": [Function],
            "onChange": [Function],
            "required": true,
            "style": Object {
              "height": 70,
              "width": 250,
            },
            "value": "some-text-value",
          },
          "ref": null,
          "rendered": null,
          "type": "textarea",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "some-text-label",
            "className": "label",
          },
          "ref": null,
          "rendered": "some-text-label",
          "type": "label",
        },
        null,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`TextArea should match snapshot when in failed state 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TextArea
    height={70}
    label="some-text-label"
    onChange={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "failed": true,
              "value": "new-text-input",
            },
          ],
        ],
      }
    }
    validator={[Function]}
    value="some-text-value"
    width={250}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <textarea
          onBlur={[Function]}
          onChange={[Function]}
          required={true}
          style={
            Object {
              "height": 70,
              "width": 250,
            }
          }
          value="new-text-input"
        />,
        <label
          className="label"
        >
          some-text-label
        </label>,
        <label
          className="error"
        >
          error
        </label>,
      ],
      "className": "text-area flex-vertical",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "onBlur": [Function],
          "onChange": [Function],
          "required": true,
          "style": Object {
            "height": 70,
            "width": 250,
          },
          "value": "new-text-input",
        },
        "ref": null,
        "rendered": null,
        "type": "textarea",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "some-text-label",
          "className": "label",
        },
        "ref": null,
        "rendered": "some-text-label",
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "error",
          "className": "error",
        },
        "ref": null,
        "rendered": "error",
        "type": "label",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <textarea
            onBlur={[Function]}
            onChange={[Function]}
            required={true}
            style={
              Object {
                "height": 70,
                "width": 250,
              }
            }
            value="new-text-input"
          />,
          <label
            className="label"
          >
            some-text-label
          </label>,
          <label
            className="error"
          >
            error
          </label>,
        ],
        "className": "text-area flex-vertical",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "onBlur": [Function],
            "onChange": [Function],
            "required": true,
            "style": Object {
              "height": 70,
              "width": 250,
            },
            "value": "new-text-input",
          },
          "ref": null,
          "rendered": null,
          "type": "textarea",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "some-text-label",
            "className": "label",
          },
          "ref": null,
          "rendered": "some-text-label",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "error",
            "className": "error",
          },
          "ref": null,
          "rendered": "error",
          "type": "label",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
